### PR TITLE
BUG/DEPR: combine dtype fixes

### LINKED ISSUE #7630: BUG: tz lost on combine first/fillna
http://stackoverflow.com/questions/24502838/pandas-tzinfo-lost-by-combine-first

```
In [10]: df1 = pd.DataFrame(columns=['UTCdatetime','abc'],
   ....:                   data=pd.to_datetime('20100101 01:01').tz_localize('UTC'),
   ....:                   index=pd.date_range('20140627', periods=1))

In [11]: print(df1)
                          UTCdatetime                        abc
2014-06-27  2010-01-01 01:01:00+00:00  2010-01-01 01:01:00+00:00

In [12]: df2 = pd.DataFrame(columns=['UTCdatetime','xyz'],
   ....:                   data=pd.to_datetime('20121212 12:12').tz_localize('UTC'),
   ....:                   index=pd.date_range('20140628', periods=1))

In [13]: 

In [13]: df1
Out[13]: 
                          UTCdatetime                        abc
2014-06-27  2010-01-01 01:01:00+00:00  2010-01-01 01:01:00+00:00

In [14]: df2
Out[14]: 
                          UTCdatetime                        xyz
2014-06-28  2012-12-12 12:12:00+00:00  2012-12-12 12:12:00+00:00

In [15]: df1.dtypes
Out[15]: 
UTCdatetime    object
abc            object
dtype: object

In [16]: df2.dtypes
Out[16]: 
UTCdatetime    object
xyz            object
dtype: object

In [17]: df2[['UTCdatetime']].combine_first(df1)
Out[17]: 
                   UTCdatetime                        abc
2014-06-27 2010-01-01 01:01:00  2010-01-01 01:01:00+00:00
2014-06-28 2012-12-12 12:12:00                        NaN

In [18]: df2[['UTCdatetime']].combine_first(df1).dtypes
Out[18]: 
UTCdatetime    datetime64[ns]
abc                    object
dtype: object
```

### LINKED ISSUE #10567: DataFrame combine_first() loses timezone information for datetime columns
xref addl example in #13650 

combine_first() loses timezone information for datetime columns

```
dts1 = pd.date_range('20150101','20150105',tz='UTC')
df1 = pd.DataFrame({'DATE':dts1})
dts2 = pd.date_range('20150103','20150105',tz='UTC')
df2 = pd.DataFrame({'DATE':dts2})
df = df1.combine_first(df2)
df.DATE[0].tz # this shows up as None
```

### LINKED ISSUE #10734: DOC: DataFrame.combine propagates nan values
The documentation for DataFrame.combine claims that the method "do[es] not propagate NaN values, so if for a (column, time) one frame is missing a value, it will default to the other frame’s value". However, this does not seem to correspond to the actual behaviour of DataFrame.combine.

Sample code:
```python
>>> import pandas as pd
>>> from operator import add
>>> a = pd.DataFrame({
...        'a': pd.Series([1, 3], index=[0, 1]),
...        'b': pd.Series([2, 3], index=[1, 2]),
...    })
>>> b = pd.DataFrame({
...         'a': pd.Series([3, 5], index=[1, 2]),
...         'c': pd.Series([1, 2, 3], index=[2, 3, 4])
...     })```
>>> a.combine(b, add)
    a   b   c
0 NaN NaN NaN
1   6 NaN NaN
2 NaN NaN NaN
3 NaN NaN NaN
4 NaN NaN NaN
```

In many cases (as in this one), it may be remedied by using the *fill_value* parameter of DataFrame.combine. However, it might be a problem when there is no acceptable neutral element for the given combining function.

```python
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Windows
OS-release: 8
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.16.2
nose: 1.3.7
Cython: 0.22
numpy: 1.9.2
scipy: 0.16.0
statsmodels: 0.6.1
IPython: 3.1.0
sphinx: 1.2.3
patsy: 0.3.0
dateutil: 2.4.2
pytz: 2015.4
bottleneck: None
tables: 3.1.1
numexpr: 2.3.1
matplotlib: 1.4.3
openpyxl: 1.8.5
xlrd: 0.9.3
xlwt: None
xlsxwriter: 0.6.7
lxml: 3.4.2
bs4: 4.3.2
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.6
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
```

### LINKED ISSUE #13947: Unified common dtype discovery
Common dtype discovery (like `np.find_common_type`) should be unified into an internal function, with proper handling of `pandas` dtypes. As of #13917, there are two such implementations: 

1. `pandas/types/common.py:_lcd_dtypes`, which works on `numpy` dtypes only, but differently than `np.find_common_type`.
2. `pandas/types/cast.py:_find_common_type`, which just uses `np.find_common_type` internally. 

Exchanging the first one with the second (which is, judging by the code, more proper), breaks some tests. 

Proposed tasks:
- [ ] Convert usages of `_lcd_dtypes` into `_find_common_types`.
- [ ] Extend `_find_common_types` to properly evaluate `pandas` dtypes.    

Occurences of `_lcd_dtypes`, the one in `internals.py` is removed by #13917:
```
[jreback-~/pandas] grin _lcd_dtype pandas
pandas/core/frame.py:
   47 :                                  _lcd_dtypes,
 3705 :                 new_dtype = _lcd_dtypes(this_dtype, other_dtype)
pandas/core/internals.py:
 4438 :     def _lcd_dtype(l):
 4473 :         lcd = _lcd_dtype(counts[IntBlock])
 4489 :         return _lcd_dtype(counts[FloatBlock] + counts[SparseBlock])
pandas/types/common.py:
  389 : def _lcd_dtypes(a_dtype, b_dtype):
```

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] closes #7630, closes #10567 (fix `combine_first` dtype)
- [x] closes #13947 (`_find_common_type` to support extension types)
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry
- `_find_common_type` also needs to handle numpy datetimelike to be compat with pandas because:

```
np.find_common_type([np.dtype('datetime64[ns]'), np.dtype('datetime64[ns]')], [])
# dtype('O')
```
- another `comibine_first` related issue  #7509 is NOT fixed because it is related to filling with `align`

### CODE REVIEW

File: pandas/tests/types/test_cast.py [hunk 1]
```diff
@@ -218,18 +218,37 @@ def test_numpy_dtypes(self):
...
-                             CategoricalDtype)
-        with self.assertRaises(TypeError):
-            self.assertEqual(_find_common_type([DatetimeTZDtype()]),
-                             DatetimeTZDtype)
+        dtype = CategoricalDtype()
+        self.assertEqual(_find_common_type([dtype]), 'category')
+        self.assertEqual(_find_common_type([dtype, dtype]), 'category')
+        self.assertEqual(_find_common_type([np.object, dtype]), np.object)
+
+        dtype = DatetimeTZDtype(unit='ns', tz='US/Eastern')
```
Comment (jreback) [inline pandas/tests/types/test_cast.py]: add a test for a datetime64[ns] w/tz against datetime64[ns]. (and maybe with np.object)

File: pandas/types/cast.py [hunk 2]
```diff
@@ -866,8 +866,23 @@ def _possibly_cast_to_datetime(value, dtype, errors='raise'):
...
 def _find_common_type(types):
     """Find a common data type among the given dtypes."""
-    # TODO: enable using pandas-specific types
+
+    if len(types) == 0:
+        raise ValueError('no types given')
+
+    first = types[0]
+    # workaround for find_common_type([np.dtype('datetime64[ns]')] * 2)
+    # => object
```
Comment (jreback) [inline pandas/types/cast.py]: this function can *almost* get rid of: https://github.com/pydata/pandas/blob/master/pandas/core/internals.py#L4799

though we use a bit more 'generic' types there

File: pandas/types/common.py [hunk 3]
```diff
@@ -386,33 +386,6 @@ def _validate_date_like_dtype(dtype):
                          (dtype.name, dtype.type.__name__))
 
 
-def _lcd_dtypes(a_dtype, b_dtype):
-    """ return the lcd dtype to hold these types """
-
```
Comment (jreback) [inline pandas/types/common.py]: excellent!

File: pandas/tests/frame/test_combine_concat.py [hunk 4]
```diff
@@ -335,6 +335,148 @@ def test_combine_first_mixed_bug(self):
         df2 = df0.combine_first(df1)
         assert_frame_equal(df2, df0)
 
+    def test_combine_first_align_nan(self):
```
Comment (jreback) [inline pandas/tests/frame/test_combine_concat.py]: maybe make separate TestCombineFirst

File: pandas/core/frame.py [hunk 5]
```diff
@@ -3700,17 +3709,20 @@ def combine(self, other, func, fill_value=None, overwrite=True):
                 otherSeries[other_mask] = fill_value
 
             # if we have different dtypes, possibily promote
-            new_dtype = this_dtype
-            if this_dtype != other_dtype:
-                new_dtype = _lcd_dtypes(this_dtype, other_dtype)
-                series = series.astype(new_dtype)
+            if notnull(series).all():
+                new_dtype = this_dtype
```
Comment (jreback) [inline pandas/core/frame.py]: in another life ``.combine`` functionaility could be pushed to block-manager (there *is* a combine there already, but it has a different purpose)

**Comment by jreback** [comment]
xref to #4094 (though that's another story)

**Comment by jreback** [comment]
very minor change. lgtm otherwise. ping when green.

**Comment by jorisvandenbossche** [comment]
Not really sure we should actually deprecate `DataFramecombine`. Although I have never used it, it does serve a purpose that is not possible to achieve with "the flexible arithmetic methods like `add`" when you pass it a custom function (unlike the deprecated `combineAdd`, which definitely can be done with `add`)

**Comment by sinhrks** [comment]
I've commented on the original.

**Comment by jreback** [comment]
ok, so as discussed in the issue. let's take out the combine deprecation from this PR. ping when ready.

**Comment by sinhrks** [comment]
Exclude ``combine`` deprecation, and now all green.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**